router-tests/headers_test.go

```diff
@@ -227,3 +228,230 @@ func TestForwardHeaders(t *testing.T) {
+		t.Parallel()
+
+		testenv.Run(t, &testenv.Config{
+			ModifyEngineExecutionConfiguration: func(cfg *config.EngineExecutionConfiguration) {
+				cfg.WebSocketReadTimeout = time.Millisecond * 10
+			},
+			RouterOptions: []core.Option{
+				core.WithHeaderRules(headerRules),
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+				Query:      `query { initialPayload }`,
+				Extensions: []byte(`{"token":"123"}`),
+			})
+			require.Equal(t, `{"data":{"initialPayload":{"extensions":{"token":"123"},"query":"{initialPayload}"}}}`, res.Body)
+		})
+	})
+
+	t.Run("ws", func(t *testing.T) {
+		t.Parallel()
+
+		cases := []struct {
+			headerName   string
+			headerRename string
+			payload      string
+			testName     string
+		}{
+			{headerNameInSubgraphRule, headerRenameInSubgraphRule, subscriptionForSubgraphRulePayload, "subgraph rule"},
+		}
+		testenv.Run(t, &testenv.Config{
+			ModifyEngineExecutionConfiguration: func(cfg *config.EngineExecutionConfiguration) {
+				cfg.WebSocketReadTimeout = time.Millisecond * 10
+			},
+			RouterOptions: []core.Option{
+				core.WithHeaderRules(headerRules),
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			for _, c := range cases {
+				c := c
+				t.Run(c.testName, func(t *testing.T) {
+
+					header := http.Header{
+						c.headerName: []string{headerValue},
+					}
+					conn := xEnv.InitGraphQLWebSocketConnection(header, nil)
+					err := conn.WriteJSON(&testenv.WebSocketMessage{
+						ID:      "1",
+						Type:    "subscribe",
+						Payload: []byte(c.payload),
+					})
+					require.NoError(t, err)
+					var msg testenv.WebSocketMessage
+					err = conn.ReadJSON(&msg)
+					require.NoError(t, err)
+					require.JSONEq(t, `{"data":{"headerValue":{"value":"`+headerValue+`"}}}`, string(msg.Payload))
+				})
+			}
+		})
+	})
+
+	t.Run("ws with client extension", func(t *testing.T) {
+		t.Parallel()
+
+		testenv.Run(t, &testenv.Config{
+			ModifyEngineExecutionConfiguration: func(cfg *config.EngineExecutionConfiguration) {
+				cfg.WebSocketReadTimeout = time.Millisecond * 10
+			},
+			RouterOptions: []core.Option{
+				core.WithHeaderRules(headerRules),
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			conn := xEnv.InitGraphQLWebSocketConnection(nil, nil)
+			err := conn.WriteJSON(&testenv.WebSocketMessage{
+				ID:      "1",
+				Type:    "subscribe",
+				Payload: []byte(`{"query":"subscription { headerValue(name:\"light\", repeat:3) { value initialPayload }}","extensions":{"token":"123"}}`),
+			})
+			require.NoError(t, err)
+			var msg testenv.WebSocketMessage
+			err = conn.ReadJSON(&msg)
+			require.NoError(t, err)
+			require.Equal(t, `{"data":{"headerValue":{"value":"","initialPayload":{"extensions":{"token":"123","upgradeHeaders":{"User-Agent":["Go-http-client/1.1"]}}}}}}`, string(msg.Payload))
+		})
+	})
+
+	t.Run("ws with multiple conns", func(t *testing.T) {
```

Comment:
Why do you think multiple connections makes a difference?

Reply:
I copied and modified the TestForwardHeaders code.